libkodimote/mpris2/mprisplayer.cpp

```diff
@@ -0,0 +1,420 @@
+        return QString("/org/mpris/MediaPlayer2/Track/channel/%1").arg(item->channelId());
+    } else if (item->songId() != -1) {
+        return QString("/org/mpris/MediaPlayer2/Track/song/%1").arg(item->songId());
+    } else if (item->recordingId() != -1) {
+        return QString("/org/mpris/MediaPlayer2/Track/recording/%1").arg(item->recordingId());
+    }
+
+    return "";
+}
+
+void MprisPlayer::sendPropertyChanged(const QString &property)
+{
+    sendPropertiesChanged(QStringList(property));
+}
+
+void MprisPlayer::sendPropertiesChanged(const QStringList &properties)
+{
+    QDBusMessage signal = QDBusMessage::createSignal("/org/mpris/MediaPlayer2", "org.freedesktop.DBus.Properties", "PropertiesChanged");
+    signal << "org.mpris.MediaPlayer2.Player";
+    QVariantMap changedProps;
+    foreach (QString property, properties) {
+        changedProps.insert(property, this->property(property.toUtf8()));
+    }
+    signal << changedProps;
+    signal << QStringList();
+    QDBusConnection::sessionBus().send(signal);
+}
+
+void MprisPlayer::activePlayerChanged()
+{
+    if (m_player == Kodi::instance()->activePlayer()) {
+        return;
+    }
+
+    if (m_player) {
+        disconnect(m_player, SIGNAL(stateChanged()), this, SLOT(stateChanged()));
+        disconnect(m_player, SIGNAL(currentItemChanged()), this, SLOT(currentItemChanged()));
+        disconnect(m_player, SIGNAL(timeChanged()), this, SLOT(timeChanged()));
+        disconnect(m_player, SIGNAL(speedChanged()), this, SLOT(speedChanged()));
+        disconnect(m_player, SIGNAL(shuffleChanged()), this, SLOT(shuffleChanged()));
+        disconnect(m_player, SIGNAL(repeatChanged()), this, SLOT(repeatChanged()));
+        disconnect(m_player->playlist(), SIGNAL(countChanged()), this, SLOT(playlistChanged()));
+        if (m_player->currentItem()) {
+            disconnect(m_player->currentItem(), SIGNAL(thumbnailChanged()), this, SLOT(thumbnailChanged()));
+        }
+    }
+
+    m_player = Kodi::instance()->activePlayer();
+
+    if (m_player) {
+        connect(m_player, SIGNAL(stateChanged()), this, SLOT(stateChanged()));
+        connect(m_player, SIGNAL(currentItemChanged()), this, SLOT(currentItemChanged()));
+        connect(m_player, SIGNAL(timeChanged()), this, SLOT(timeChanged()));
+        connect(m_player, SIGNAL(speedChanged()), this, SLOT(speedChanged()));
+        connect(m_player, SIGNAL(shuffleChanged()), this, SLOT(shuffleChanged()));
+        connect(m_player, SIGNAL(repeatChanged()), this, SLOT(repeatChanged()));
+        connect(m_player->playlist(), SIGNAL(countChanged()), this, SLOT(playlistChanged()));
+    }
+
+    currentItemChanged();
+}
+
+void MprisPlayer::volumeChanged()
+{
+    sendPropertyChanged("Volume");
+}
+
+void MprisPlayer::stateChanged()
+{
+    sendPropertyChanged("PlaybackStatus");
+}
+
+void MprisPlayer::currentItemChanged()
+{
+    connect(m_player->currentItem(), SIGNAL(thumbnailChanged()), this, SLOT(thumbnailChanged()));
```

Comment:
This crashes when disconnecting from kodi. Here's a trace: http://paste.ubuntu.com/10576993/

is there a reason why you connect this one in here and not in line 370, inside the if(m_player) like all the others?